pymodelextractor/learners/observation_table_learners/generic_lstar_learner.py

```diff
@@ -0,0 +1,239 @@
+from ast import Tuple
+from pythautomata.base_types.sequence import Sequence
+from pymodelextractor.learners.observation_table_learners.generic_observation_table import GenericObservationTable
+from pymodelextractor.learners.learning_result import LearningResult
+from pymodelextractor.teachers.generic_teacher import GenericTeacher
+import time
+import multiprocessing
+from pymodelextractor.utils.time_bound_utilities import timeout
+
+lamda = Sequence()
+
+no_log = 'none'
+info_log = 'info'
+debug_log = 'debug'
+trace_log = 'trace'
+
+class GenericLStarLearner:
+    def __init__(self, model_translator, max_states = None, max_query_lenght = None, max_time = None):
+        self._model_translator = model_translator
+        self._max_states = max_states
+        self._max_query_length = max_query_lenght
+        self._max_time = max_time
+        self._history = []
+
+    def _build_observation_table(self):
+        self._observation_table = GenericObservationTable()
+    
+    def _initialize_observation_table(self):
+        self._observation_table.exp = [lamda]
+        self._add_to_red(lamda)
+        for symbol in self._symbols:
+            self._add_to_blue(Sequence((symbol,)))
+
+    def _add_to_blue(self, sequence: Sequence) -> bool:
+        if not sequence in self._observation_table.blue:
+            self._observation_table.blue.add(sequence)
+            self._observation_table[sequence], surpassed_max_query_len = self._get_filled_row_for(
+                sequence)
+            
+            return surpassed_max_query_len
+            
+    def _add_to_red(self, sequence: Sequence) -> bool:
+        if sequence not in self._observation_table.red:
+            redValue, surpassed_max_query_len = self._get_filled_row_for(sequence)
+            self._observation_table[sequence] = redValue
+            self._observation_table.add_to_red(sequence, redValue)
+
+            return surpassed_max_query_len
+    
+    def _get_filled_row_for(self, sequence: Sequence):
+        required_suffixes = self._observation_table.exp
+        row = []
+
+        if (self._max_query_length is not None) and ( max(len(suffix) for suffix in required_suffixes) + len(sequence) > self._max_query_length):
```

Comment:
`self._max_query_length != -1`
line is too long, perhaps refactor right into it's own method